src/main/java/com/microsoft/jenkins/acs/ACSDeploymentContext.java

```
@@ -1,248 +1,497 @@
+                new TransitionInfo(new DeploymentChoiceCommand(), null, null));
+
+        // ACS with Kubernetes will add a security rule for the service port automatically,
+        // so no need to manually create one to enable the port access
+        commands.put(KubernetesDeploymentCommand.class,
+                new TransitionInfo(new KubernetesDeploymentCommand(), null, null));
+
+        commands.put(MarathonDeploymentCommand.class,
+                new TransitionInfo(new MarathonDeploymentCommand(), EnablePortCommand.class, null));
+        commands.put(EnablePortCommand.class,
+                new TransitionInfo(new EnablePortCommand(), null, null));
+        super.configure(
+                new JobContext(run, workspace, launcher, listener),
+                commands,
+                GetContainserServiceInfoCommand.class);
         this.setDeploymentState(DeploymentState.Running);
     }
 
-    public void cleanupConfigFile() {
-        if (this.localMarathonConfigFile != null) {
-            boolean ret = this.localMarathonConfigFile.delete();
-            if (!ret) {
-                logStatus("Cannot delete temporary Marathon configuration file: " + this.localMarathonConfigFile.getAbsolutePath());
-            }
+    private static SSHUserPrivateKey getSshCredentials(final String id) {
+        SSHUserPrivateKey creds = CredentialsMatchers.firstOrNull(
+                CredentialsProvider.lookupCredentials(
+                        BasicSSHUserPrivateKey.class,
+                        Jenkins.getInstance(),
+                        ACL.SYSTEM,
+                        Collections.<DomainRequirement>emptyList()),
+                CredentialsMatchers.withId(id));
+        return creds;
+    }
+
+    public static String getContainerServiceName(final String containerService) {
+        if (StringUtils.isBlank(containerService)) {
+            throw new IllegalArgumentException(Messages.ACSDeploymentContext_blankContainerService());
```

The `containerService` name is used in later build steps frequently so throwing an exception will fail the build immediately; while the `orchestratorType` is only for the convenience during the configuration time, when it comes to the build, we will infer the type from the Azure.

Consider the next question you raised that the `orchestratorType` may change after the configuration, we can update the `getOrchestratorType` method so that it raises exception rather than returning `null`.